src/object/srv_obj_migrate.c

```diff
@@ -2539,6 +2591,34 @@ migrate_cont_iter_cb(daos_handle_t ih, d_iov_t *key_iov,
 			break;
 	}
 
+	/* FIXME: keep container handle open until all ULTs completes. */
+	for (rc1 = 0; rc1 < 100; rc1++) {
+		dss_sleep(500 /* ms */);
+
+		mq_arg.obj_generated_ult = 0;
+		mq_arg.obj_executed_ult = 0;
+		mq_arg.generated_ult = 0;
+		mq_arg.executed_ult = 0;
+
+		rc = dss_thread_collective(migrate_check_one, &mq_arg, 0);
+		if (rc)
+			continue;
+
+		/**
+		 * The object ULT is generated by 0 xstream,
+		 * and dss_collective does not do collective on 0 xstream
+		 **/
+		mq_arg.obj_generated_ult += tls->mpt_obj_generated_ult;
+		if (mq_arg.obj_generated_ult == 0)
+			continue;
+
+		if (mq_arg.obj_generated_ult == mq_arg.obj_executed_ult &&
+		    mq_arg.generated_ult == mq_arg.executed_ult)
+			break;
+	}
+
+	ABT_mutex_free(&mq_arg.status_lock);
```

Comment:
Sorry, IMHO, this is a bit hacky. I think we should add reference count into dsc_cont_open() and dsc_cont_close(), and does not wait and do collective operation. 

@dmiter  Could we fix  this in a separate patch?

Reply:
Elaborated this solution one more time.